test/parallel/test-domain-promise.js

```diff
@@ -0,0 +1,25 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const domain = require('domain');
+const vm = require('vm');
+
+const d = domain.create();
+
```

Comment:
If I understand the promise hooks correctly above, the following should be true under the current implementation:

```javascript
const d1 = domain.create();
const d2 = domain.create();

var p;
d1.run(() => {
  p = Promise.resolve(42);
});

d2.run(() => {
  p.then((v) => {
    // running in d2
  });
});
```
An equally valid way of propagating domains would be for the callback passed to `then` to run in the domain of its parent promise. With either default behavior it is possible to have the callback run in either context (by binding it explicitly) so I don't think this is a fundamental limitation. It would probably be good to describe this design choice and document how to work around it in any documentation that is added as part of this PR.